keps/0007-command-asserts.md

```diff
@@ -0,0 +1,103 @@
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [User Stories](#user-stories)
+    - [User Story 1](#user-story-1)
+    - [User Story 2](#user-story-2)
+  - [Implementation Details](#implementation-details)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Graduation Criteria](#graduation-criteria)
+- [Implementation History](#implementation-history)
+
+## Summary
+
+By asserting CLI commands, kuttl can support many additional test scenarios that examing the internal state of an operator.
+
+## Motivation
+
+Operators and custom resources can be used to deploy complex applications. For example, KUDO provides operators for Apache Cassandra and Apache Kafka. Deploying custom resources of these operators may not always result in changed Kubernetes resources but also in changes that aren't surfaced by Kubernetes resources but are internal to the operator. Asserting these changes only works through accessing service APIs or running commands and parsing their output. kuttl should provide tools to assert such output.
+
+### Goals
+
+- Assert that commands examining the internal state of an operator succeed or fail
+- Keep eventual consistency when asserting commands
+
+### Non-Goals
+
+- Introduce "flows" to determine when to run commands. I.e. to run commands once before or after asserting resources
+
+## Proposal
+
+Introduce commands as part of `TestAssert`. These commands are run at a regular interval like resource assertions until they either succeed or a timeout of the assert is reached. If any of the commands fail, the assert is considered to have failed. A command's return value determines if a command failed or not.
+
+As a start, complex commands that parse output can be described using shell environments. At a later stage we can implement additional tools to simplify common tasks. E.g. to search with a regular expression in command output.
+
+### User Stories
+
+#### User Story 1
+
+The JVM options of a KUDO operator can be tuned using parameters. On deployment, the JVM options are logged in the pods running the application provided by the operator. To assert that a parameter change resulted in changed JVM options we need to run `kubectl logs pod-name` and search the logs for a specific string once the operator has been deployed.
+
+#### User Story 2
+
+An operator can be configured to encrypt it's API. To assert that the API endpoints are indeed encrypted we can run a specific `curl` command and check that it returns successfully once the operator has been deployed.
+
+### Implementation Details
+
+Command assertions use most fields of the existing commands for `TestStep` and follow their behavior. The `ignoreFailure`, `background` and `timeout` fields are removed as they are not needed here.
+
+An example that parses the logs of a pod:
+
+```yaml
+apiVersion: kuttl.dev/v1beta1
+type: TestAssert
+commands:
+  - command: ./parse-logs.sh name-of-pod-0 string-to-search-for
+    namespaced: true
+```
```

Comment:
```suggestion
```yaml
apiVersion: kuttl.dev/v1beta1
type: TestAssert
commands:
  - command: ./parse-logs.sh name-of-pod-0 string-to-search-for
    namespaced: true
timeout: 60
delay: 10
```
```
I would suggest an addition of an "assertDelay". This is currently fixed at 1 second, meaning the assert is checked after a delay of 1 second. 
When we add commands, we might want to increase this delay, maybe because of very expensive checks. 
An `initialDelay` might also make sense, although adding delays to check is often bad design...

Reply:
not a fan of delay... If we have one... we should be clear a delay from what... I would rather have the declarative asserts successful as a criteria of when to start commands... time delays are a common source of flaky tests :(

Reply:
Well, we currently have a fixed delay of 1 second between rechecks. I.e. when an assert fails, there is a one-second delay until it is checked again. Might be worth to make that configurable.